### PR TITLE
Add patch_reloc method for bin/coff plugin

### PR DESCRIPTION
**Pull request by pelijah**
**Your checklist for this pull request**
- [ ] I've read the [guidelines for contributing](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md) to this repository
- [ ] I made sure to follow the project's [coding style](https://github.com/radareorg/radare2/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [radare2 book](https://github.com/radareorg/radare2book) with the relevant information (if needed)

**Detailed description**
Another step on the road to resolving #7310.

Before:
```asm
[0x0000b45c]> pdf @ sym._TIFFNumberOfDirectories
/ 144: sym._TIFFNumberOfDirectories (int32_t arg_18h);
|           ; var int32_t var_ch @ esp+0x18
|           ; var int32_t var_10h @ esp+0x1c
|           ; arg int32_t arg_18h @ esp+0x24
|           0x0000b4dc      83ec08         sub esp, 8
|           0x0000b4df      53             push ebx
|           0x0000b4e0      56             push esi
|           0x0000b4e1      57             push edi
|           0x0000b4e2      8b7c2418       mov edi, dword [arg_18h]
|           0x0000b4e6      f7470c000008.  test dword [edi + 0xc], 0x80000
|       ,=< 0x0000b4ed      750a           jne 0xb4f9
|       |   0x0000b4ef      8b87ac010000   mov eax, dword [edi + 0x1ac]
|       |   0x0000b4f5      33c9           xor ecx, ecx
|      ,==< 0x0000b4f7      eb0c           jmp 0xb505
|      |`-> 0x0000b4f9      8b87b0010000   mov eax, dword [edi + 0x1b0]
|      |    0x0000b4ff      8b8fb4010000   mov ecx, dword [edi + 0x1b4]
|      |    ; CODE XREF from sym._TIFFNumberOfDirectories @ 0xb4f7
|      `--> 0x0000b505      8944240c       mov dword [var_ch], eax
|           0x0000b509      33f6           xor esi, esi
|           0x0000b50b      0bc1           or eax, ecx
|           0x0000b50d      894c2410       mov dword [var_10h], ecx
|       ,=< 0x0000b511      742d           je 0xb540
|       |   0x0000b513      bbffff0000     mov ebx, 0xffff             ; "real@ff7fffff"
|       |   0x0000b518      0f1f4000       nop dword [eax]
|      .--> 0x0000b51c      6a00           push 0
|      :|   0x0000b51e      8d442410       lea eax, [var_ch]
|      :|   0x0000b522      50             push eax
|      :|   0x0000b523      57             push edi
|      :|   0x0000b524      e800000000     call _TIFFAdvanceDirectory  ; RELOC 20 _TIFFAdvanceDirectory @ 0x00000000 + 0x0
|      :|   ; CALL XREF from sym._TIFFNumberOfDirectories @ 0xb524
|      :|   0x0000b529      83c40c         add esp, 0xc
|      :|   0x0000b52c      85c0           test eax, eax
|     ,===< 0x0000b52e      7410           je 0xb540
|     |:|   0x0000b530      663bf3         cmp si, bx
|    ,====< 0x0000b533      7415           je $SG8910
|    ||:|   0x0000b535      8b44240c       mov eax, dword [var_ch]
|    ||:|   0x0000b539      46             inc esi
|    ||:|   0x0000b53a      0b442410       or eax, dword [var_10h]
|    ||`==< 0x0000b53e      75dc           jne 0xb51c
|    |`-`-> 0x0000b540      5f             pop edi
|    |      0x0000b541      668bc6         mov ax, si
|    |      0x0000b544      5e             pop esi
|    |      0x0000b545      5b             pop ebx
|    |      0x0000b546      83c408         add esp, 8
|    |      0x0000b549      c3             ret
(reloc._SG8910)
|    `----> 0x0000b54a      6800000000     push 0                      ; RELOC 6 $SG8910 @ 0x00000000 + 0x0
(reloc._module__1__TIFFNumberOfDirectories__9_9)
|           0x0000b54f      6800000000     push 0                      ; RELOC 6 ?module@?1??TIFFNumberOfDirectories@@9@9 @ 0x00000000 + 0x0
|           0x0000b554      ffb774020000   push dword [edi + 0x274]
|           0x0000b55a      e800000000     call _TIFFErrorExt          ; RELOC 20 _TIFFErrorExt @ 0x00000000 + 0x0
|           ; CALL XREF from sym._TIFFNumberOfDirectories @ 0xb55a
|           0x0000b55f      83c40c         add esp, 0xc
|           0x0000b562      668bc3         mov ax, bx
|           0x0000b565      5f             pop edi
|           0x0000b566      5e             pop esi
|           0x0000b567      5b             pop ebx
|           0x0000b568      83c408         add esp, 8
\           0x0000b56b      c3             ret
```
After:
```asm
[0x0000b45c]> pdf @ sym._TIFFNumberOfDirectories
/ 144: sym._TIFFNumberOfDirectories (int32_t arg_18h);
|           ; var int32_t var_ch @ esp+0x18
|           ; var int32_t var_10h @ esp+0x1c
|           ; arg int32_t arg_18h @ esp+0x24
|           0x0000b4dc      83ec08         sub esp, 8
|           0x0000b4df      53             push ebx
|           0x0000b4e0      56             push esi
|           0x0000b4e1      57             push edi
|           0x0000b4e2      8b7c2418       mov edi, dword [arg_18h]
|           0x0000b4e6      f7470c000008.  test dword [edi + 0xc], 0x80000
|       ,=< 0x0000b4ed      750a           jne 0xb4f9
|       |   0x0000b4ef      8b87ac010000   mov eax, dword [edi + 0x1ac]
|       |   0x0000b4f5      33c9           xor ecx, ecx
|      ,==< 0x0000b4f7      eb0c           jmp 0xb505
|      |`-> 0x0000b4f9      8b87b0010000   mov eax, dword [edi + 0x1b0]
|      |    0x0000b4ff      8b8fb4010000   mov ecx, dword [edi + 0x1b4]
|      |    ; CODE XREF from sym._TIFFNumberOfDirectories @ 0xb4f7
|      `--> 0x0000b505      8944240c       mov dword [var_ch], eax
|           0x0000b509      33f6           xor esi, esi
|           0x0000b50b      0bc1           or eax, ecx
|           0x0000b50d      894c2410       mov dword [var_10h], ecx
|       ,=< 0x0000b511      742d           je 0xb540
|       |   0x0000b513      bbffff0000     mov ebx, 0xffff             ; "real@ff7fffff"
|       |   0x0000b518      0f1f4000       nop dword [eax]
|      .--> 0x0000b51c      6a00           push 0
|      :|   0x0000b51e      8d442410       lea eax, [var_ch]
|      :|   0x0000b522      50             push eax
|      :|   0x0000b523      57             push edi
|      :|   0x0000b524      e823230000     call sym._TIFFAdvanceDirectory; RELOC 32 _TIFFAdvanceDirectory @ 0x0000d84c + 0x0
|      :|   0x0000b529      83c40c         add esp, 0xc
|      :|   0x0000b52c      85c0           test eax, eax
|     ,===< 0x0000b52e      7410           je 0xb540
|     |:|   0x0000b530      663bf3         cmp si, bx
|    ,====< 0x0000b533      7415           je $SG8910
|    ||:|   0x0000b535      8b44240c       mov eax, dword [var_ch]
|    ||:|   0x0000b539      46             inc esi
|    ||:|   0x0000b53a      0b442410       or eax, dword [var_10h]
|    ||`==< 0x0000b53e      75dc           jne 0xb51c
|    |`-`-> 0x0000b540      5f             pop edi
|    |      0x0000b541      668bc6         mov ax, si
|    |      0x0000b544      5e             pop esi
|    |      0x0000b545      5b             pop ebx
|    |      0x0000b546      83c408         add esp, 8
|    |      0x0000b549      c3             ret
(reloc._SG8910)
|    `----> 0x0000b54a      68f3b00000     push sym._SG8910            ; 0xb0f3 ; "Directory count exceeded 65535 limit, giving up on counting."; RELOC 32 $SG8910 @ 0x0000b0f3 + 0x0
(reloc._module__1__TIFFNumberOfDirectories__9_9)
|           0x0000b54f      6830b40000     push sym._module__1__TIFFNumberOfDirectories__9_9 ; 0xb430 ; "TIFFNumberOfDirectories"; RELOC 32 ?module@?1??TIFFNumberOfDirectories@@9@9 @ 0x0000b430 + 0x0
|           0x0000b554      ffb774020000   push dword [edi + 0x274]
|           0x0000b55a      e8b64a0000     call _TIFFErrorExt
|           0x0000b55f      83c40c         add esp, 0xc
|           0x0000b562      668bc3         mov ax, bx
|           0x0000b565      5f             pop edi
|           0x0000b566      5e             pop esi
|           0x0000b567      5b             pop ebx
|           0x0000b568      83c408         add esp, 8
\           0x0000b56b      c3             ret
```

**Test plan**

...

**Closing issues**

...

### CODE REVIEW

**Comment by pelijah** [comment]
@ret2libc because imported symbols aren't physically presented in COFF object file. They are just name containers for relocs which are patched during linking of executable. As you can see here call of imported symbol just creates xref to the next instruction. 
 ```asm
0x0002310e      e800000000     call _errno                 ; RELOC 0 _errno
; CALL XREF from sym._common_fseek__YAHV__crt_stdio_stream___JH_Z @ 0x2310e
0x00023113      c70016000000   mov dword [rax], 0x16       ; [0x16:4]=0x74636572
```

**Comment by pelijah** [comment]
And noreturn propagation doesn't work in case of imports
```asm
/ 50: entry0 (int64_t arg1, int64_t arg_61h);
|           ; arg int64_t arg_61h @ rbp+0x61
|           ; arg int64_t arg1 @ rdi
|           0x00000064      41b900000000   mov r9d, 0                  ; [00] -r-x section size 39 named .text_0
|           0x0000006a      4c8d051a0000.  lea r8, [0x0000008b]        ; sym..data ; "Win64 assembly"; RELOC 32 .data @ 0x00000000 + 0x0
(reloc..data)
|           0x00000071      488d15130000.  lea rdx, [0x0000008b]       ; sym..data ; "Win64 assembly"; RELOC 32 .data @ 0x00000000 + 0x0
|           0x00000078      48c7c1000000.  mov rcx, 0
(reloc.MessageBoxA)
|           0x0000007f      e800000000     call MessageBoxA            ; RELOC 4 MessageBoxA
|           ; CALL XREF from entry0 @ 0x7f
|           0x00000084      89c1           mov ecx, eax
(reloc.ExitProcess)
|           0x00000086      e800000000     call ExitProcess            ; RELOC 4 ExitProcess
|           ; CALL XREFS from entry0 @ 0x6a, 0x71, 0x86
|           ;-- str.Win64_assembly:
|           ;-- section..data_1:
|           ;-- .data:
|           0x0000008b     .string "Win64 assembly" ; len=15           ; arg1 ; [01] -rw- section size 28 named .data_1
\           0x00000095     .string "Win64 assembly" ; len=15
```

**Comment by pelijah** [comment]
```
Error while parsing command: `pd 1 @ reloc._TIFFVGetField - 1`
```
So is it illegal to use `pd 1 @ reloc._TIFFVGetField - 1`?

**Comment by ret2libc** [comment]
In the newshell you should not have spaces, otherwise the reloc.xxx - 1 will be interpreted as 3 args: reloc.xxx, -, 1. Since @ expect only one argument after it, that is considered a parsing error, at least in the current state of things

**Comment by ret2libc** [comment]
BTW @pelijah any comment input on this thing is appreciated. Do you think it is wrong to not support spaces between operations?

**Comment by pelijah** [comment]
> 
> 
> BTW @pelijah any comment input on this thing is appreciated. Do you think it is wrong to not support spaces between operations?

I am the last person you should ask about this.

File: libr/bin/p/bin_coff.c [hunk 1]
```diff
@@ -113,19 +113,24 @@ static bool _fill_bin_symbol(RBin *rbin, struct r_bin_coff_obj *bin, int idx, RB
...
+
 static RBinImport *_fill_bin_import(struct r_bin_coff_obj *bin, int idx) {
 	RBinImport *ptr = R_NEW0 (RBinImport);
 	if (!ptr || idx < 0 || idx > bin->hdr.f_nsyms) {
 		free (ptr);
 		return NULL;
 	}
 	struct coff_symbol *s = &bin->symbols[idx];
-	if (s->n_scnum != COFF_SYM_SCNUM_UNDEF || s->n_sclass != COFF_SYM_CLASS_EXTERNAL) {
+	if (!is_imported_symbol(s)) {
```
Comment (trufae) [inline libr/bin/p/bin_coff.c:133]: missing space before (

File: libr/bin/p/bin_coff.c [hunk 2]
```diff
@@ -113,19 +113,24 @@ static bool _fill_bin_symbol(RBin *rbin, struct r_bin_coff_obj *bin, int idx, RB
 		ptr->type = r_str_constpool_get (&rbin->constpool, sdb_fmt ("%i", s->n_sclass));
 		break;
 	}
+	ptr->vaddr = ptr->paddr;
 	ptr->size = 4;
 	ptr->ordinal = 0;
 	return true;
 }
 
+static bool is_imported_symbol(struct coff_symbol *s) {
```
Comment (trufae) [inline libr/bin/p/bin_coff.c:122]: inline

File: libr/bin/p/bin_coff.c [hunk 3]
```diff
@@ -255,63 +261,171 @@ static RList *libs(RBinFile *bf) {
...
+#define BYTES_PER_IMP_RELOC		8
+
+static RList *_relocs_list(RBin *rbin, struct r_bin_coff_obj *bin, bool patch, ut64 impmap) {
 	RBinReloc *reloc;
 	struct coff_reloc *rel;
 	int j, i = 0;
 	RList *list_rel;
 	list_rel = r_list_new ();
-	if (!list_rel || !bin || !bin->scn_hdrs) {
+	HtUU *imp_vaddr_ht = NULL;
```
Comment (trufae) [inline libr/bin/p/bin_coff.c:272]: use ternary operator instead of 3 extra lines

File: libr/bin/p/bin_coff.c [hunk 4]
```diff
@@ -255,63 +261,171 @@ static RList *libs(RBinFile *bf) {
...
+	if (patch) {
+		imp_vaddr_ht = ht_uu_new0 ();
+	}
+	if (!list_rel || !bin || !bin->scn_hdrs || (patch && !imp_vaddr_ht)) {
```
Comment (trufae) [inline libr/bin/p/bin_coff.c:276]: use r_return_val_if_fail.

File: libr/bin/p/bin_coff.c (hunk 3 above)
Comment (trufae) [reply libr/bin/p/bin_coff.c:272]: move this initializers after the checks below to avoid a leak

File: libr/bin/p/bin_coff.c (hunk 2 above)
Comment (ret2libc) [reply libr/bin/p/bin_coff.c:122]: not sure it's really necessary... the compiler will inline this anyway if possible, when you compile in release mode.

File: libr/bin/p/bin_coff.c (hunk 4 above)
Comment (ret2libc) [reply libr/bin/p/bin_coff.c:276]: I think only `bin` can be placed in a `r_return_val_if_fail`. `list_rel` and `imp_vaddr_ht` are allocated at runtime and should not be checked with `r_return_`.

**Pull request merged**